dialogue-core/src/main/java/com/palantir/dialogue/core/RequestSizeMetricsChannel.java

```diff
@@ -0,0 +1,144 @@
+import com.palantir.conjure.java.client.config.ClientConfiguration;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.EndpointChannel;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.RequestBody;
+import com.palantir.dialogue.Response;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Optional;
+
+final class RequestSizeMetricsChannel implements EndpointChannel {
+    private static final SafeLogger log = SafeLoggerFactory.get(RequestSizeMetricsChannel.class);
+    private final EndpointChannel delegate;
+    private final Histogram requestSize;
+
+    static EndpointChannel create(Config cf, EndpointChannel channel, Endpoint endpoint) {
+        ClientConfiguration clientConf = cf.clientConf();
+        return new RequestSizeMetricsChannel(channel, endpoint, clientConf.taggedMetricRegistry());
+    }
+
+    RequestSizeMetricsChannel(EndpointChannel delegate, Endpoint endpoint, TaggedMetricRegistry registry) {
+        this.delegate = delegate;
+        DialogueClientMetrics dialogueClientMetrics = DialogueClientMetrics.of(registry);
+        this.requestSize = dialogueClientMetrics
+                .requestsSize()
+                .serviceName(endpoint.serviceName())
+                .endpoint(endpoint.endpointName())
+                .build();
+    }
+
+    @Override
+    public ListenableFuture<Response> execute(Request request) {
+        Request augmentedRequest = wrap(request);
+        return delegate.execute(augmentedRequest);
+    }
+
+    private Request wrap(Request request) {
+        Optional<RequestBody> body = request.body();
+        if (body.isEmpty()) {
+            // No need to record empty bodies
+            return request;
+        }
+
+        return Request.builder()
+                .from(request)
+                .body(new RequestSizeRecordingRequestBody(body.get(), this.requestSize))
+                .build();
+    }
+
+    private class RequestSizeRecordingRequestBody implements RequestBody {
+        private final RequestBody delegate;
+        private final Histogram size;
+        private SizeTrackingOutputStream out;
+
+        RequestSizeRecordingRequestBody(RequestBody requestBody, Histogram size) {
+            this.delegate = requestBody;
+            this.size = size;
+            // we'll never actually write to this output stream, but is safe to perform all operations on in case a
+            // client closes without calling write.
+            this.out = new SizeTrackingOutputStream(OutputStream.nullOutputStream(), size);
+        }
+
+        @Override
+        public void writeTo(OutputStream output) throws IOException {
+            out = new SizeTrackingOutputStream(output, size);
+            delegate.writeTo(out);
+        }
```

Comment:
Instead of maintaining a field, what do you think about something along these lines?

```suggestion
        @Override
        public void writeTo(OutputStream output) throws IOException {
            CountingOutputStream countingOut = new CountingOutputStream(out)
            delegate.writeTo(countingOut);
            size.update(countingOut.getCount());
        }
```

This way, if the delegate calls close on `countingOut`, it is passed through to `out`, however we don't require close to be called on the wrapper stream here. Keeps everything nicely contained in this method.

The guava [CountingOutputStream](https://github.com/google/guava/blob/master/guava/src/com/google/common/io/CountingOutputStream.java) handles counting bytes for us :-]